model/common/src/icon4py/model/common/orchestration/halo_exchange.py

```diff
@@ -0,0 +1,325 @@
+        dim: dimension over which the exchange is performed
+        unique_id:
+        wait: Async exchange or not
+        counter: counter to define the global variables only once
+    """
+    tasklet = dace.sdfg.nodes.Tasklet(
+        "_halo_exchange_",
+        inputs=None,
+        outputs=None,
+        code="",
+        language=dace.dtypes.Language.CPP,
+        side_effects=True,
+    )
+    state.add_node(tasklet)
+
+    in_connectors = {}
+    out_connectors = {}
+
+    if counter == 0:  # define them only once
+        for buffer_name in ghex_ptr_names:
+            sdfg.add_scalar(buffer_name, dtype=dace.uintp)
+            buffer = state.add_read(buffer_name)
+            in_connectors["IN_" + buffer_name] = dace.uintp.dtype
+            state.add_edge(
+                buffer, None, tasklet, "IN_" + buffer_name, Memlet(buffer_name, subset="0")
+            )
+
+    for i, (buffer_name, data_descriptor) in enumerate(global_buffers.items()):
+        sdfg.add_array(
+            buffer_name,
+            data_descriptor.shape,
+            data_descriptor.dtype,
+            storage=data_descriptor.storage,
+            strides=data_descriptor.strides,
+        )
+        buffer = state.add_read(buffer_name)
+        in_connectors["IN_" + f"field_{i}"] = dtypes.pointer(data_descriptor.dtype)
+        state.add_edge(
+            buffer,
+            None,
+            tasklet,
+            "IN_" + f"field_{i}",
+            Memlet.from_array(buffer_name, data_descriptor),
+        )
+
+        update = state.add_write(buffer_name)
+        out_connectors["OUT_" + f"field_{i}"] = dtypes.pointer(data_descriptor.dtype)
+        state.add_edge(
+            tasklet,
+            "OUT_" + f"field_{i}",
+            update,
+            None,
+            Memlet.from_array(buffer_name, data_descriptor),
+        )
+
+    # Dummy return: preserve same interface with non-DaCe version
+    sdfg.add_scalar(name="__return", dtype=dace.int32)
+    ret = state.add_write("__return")
+    state.add_edge(tasklet, "__out", ret, None, dace.Memlet(data="__return", subset="0"))
+    out_connectors["__out"] = dace.int32
+
+    tasklet.in_connectors = in_connectors
+    tasklet.out_connectors = out_connectors
+    tasklet.environments = [f"{DaceGHEX.__module__}.{DaceGHEX.__name__}"]
+
+    pattern_type = exchange._patterns[dim].__cpp_type__
+    domain_descriptor_type = exchange._domain_descriptors[dim].__cpp_type__
+    communication_object_type = exchange._comm.__cpp_type__
+    communication_handle_type = communication_object_type[
+        communication_object_type.find("<") + 1 : communication_object_type.rfind(">")
+    ]
+
+    # Tasklet code generation part
+    fields_desc_glob_vars = "\n"
+    fields_desc = """
+    bool levels_first;
+    std::size_t outer_strides;
+    std::size_t levels;
+
+    """
+    descr_unique_names = []
+    for i, arg in enumerate(copy.deepcopy(list(global_buffers.values()))):
```

Comment:
Why is a deepcopy needed here? Is the loop body modifying the nested contents of the sequence?